### PR TITLE
R4R: Refactor Total Supply

### PR DESCRIPTION
**Pull request by zhangyelong**
Resolves: #1603 
Blocks: #1673

### CODE REVIEW

File: app/v1/auth/keeper.go [hunk 1]
```diff
@@ -271,6 +272,88 @@ func (am AccountKeeper) DecreaseTotalLoosenToken(ctx sdk.Context, coins sdk.Coin
 		"decreaseCoins", decreaseCoins.String(), "totalLoosenToken", totalLoosenToken.String())
 }
 
+// Turn a token demom to key used to get it from the account store
+func TotalSupplyStoreKey(denom string) []byte {
+	return append(totalSupplyKeyPrefix, []byte(denom)...)
```
Comment (chengwenxi) [inline app/v1/auth/keeper.go]: Why not token_id ?

File: app/v1/auth/keeper.go [hunk 2]
```diff
@@ -271,6 +272,88 @@ func (am AccountKeeper) DecreaseTotalLoosenToken(ctx sdk.Context, coins sdk.Coin
...
+}
+
+func (am AccountKeeper) IncreaseTotalSupply(ctx sdk.Context, coin sdk.Coin) sdk.Error {
+	// parameter checking
+	if coin == (sdk.Coin{}) || !coin.IsPositive() || coin.IsZero() {
```
Comment (chengwenxi) [inline app/v1/auth/keeper.go]: !coin.IsPositive() means <0 && ==0

File: app/v1/auth/keeper.go [hunk 3]
```diff
@@ -271,6 +272,88 @@ func (am AccountKeeper) DecreaseTotalLoosenToken(ctx sdk.Context, coins sdk.Coin
...
+
+	ctx.Logger().Info("Execute IncreaseTotalSupply Succeeded",
+		"increaseCoins", coin.String(), "totalSupply", totalSupply.String())
+
+	return nil
+}
+
+func (am AccountKeeper) DecreaseTotalSupply(ctx sdk.Context, coin sdk.Coin) sdk.Error {
+	// parameter checking
+	if coin == (sdk.Coin{}) || !coin.IsPositive() || coin.IsZero() {
```
Comment (chengwenxi) [inline app/v1/auth/keeper.go]: ++

File: app/v1/auth/keeper.go (hunk 1 above)
Comment (zhangyelong) [reply app/v1/auth/keeper.go]: In system layer, all operations use the min symbol of tokens just like `iris-atto`, denom is the unique identity of a token, which doesn't need more convert operations, I don't think token_id is a better choice :)

Comment (chengwenxi) [reply app/v1/auth/keeper.go]: You are right.

**Comment by zhangyelong** [comment]
Wait a minute, bank cli tests need to be fixed

**Comment by zhangyelong** [comment]
> Wait a minute, bank cli tests needs to be fixed

Done

File: app/v1/auth/keeper.go (hunk 1 above)
Comment (haifengxi) [reply app/v1/auth/keeper.go]: The same `token_id` that has been used to construct the key to store an asset might be a better choice -- reusable code `GetKeyIDFromUniqueID`, more consistency, and support for prefix search (although we might not need it for total supply, it's still good to have).

Comment (haifengxi) [inline app/v1/auth/keeper.go]: The same `token_id` that has been used to construct the key to store an asset might be a better choice -- reusable code `GetKeyIDFromUniqueID`, more consistency, and support for prefix search (although we might not need it for total supply, it's still good to have).

File: app/v1/asset/handler.go [hunk 4]
```diff
@@ -30,9 +30,8 @@ func handleIssueToken(ctx sdk.Context, k Keeper, msg MsgIssueToken) sdk.Result {
 	var token FungibleToken
 	switch msg.Family {
 	case FUNGIBLE:
-		totalSupply := msg.InitialSupply
 		decimal := int(msg.Decimal)
-		token = NewFungibleToken(msg.Source, msg.Gateway, msg.Symbol, msg.Name, msg.Decimal, msg.SymbolAtSource, msg.SymbolMinAlias, sdk.NewIntWithDecimal(int64(msg.InitialSupply), decimal), sdk.NewIntWithDecimal(int64(totalSupply), decimal), sdk.NewIntWithDecimal(int64(msg.MaxSupply), decimal), msg.Mintable, msg.Owner)
+		token = NewFungibleToken(msg.Source, msg.Gateway, msg.Symbol, msg.Name, msg.Decimal, msg.SymbolAtSource, msg.SymbolMinAlias, sdk.NewIntWithDecimal(int64(msg.InitialSupply), decimal), sdk.NewIntWithDecimal(int64(msg.MaxSupply), decimal), msg.Mintable, msg.Owner)
```
Comment (secret2830) [inline app/v1/asset/handler.go]: Now that `supply` is an Int value, conversion to int64 will cause precision loss

Comment (zhangyelong) [reply app/v1/asset/handler.go]: No for uint64 to int64

File: app/v1/auth/keeper.go (hunk 1 above)
Comment (zhangyelong) [reply app/v1/auth/keeper.go]: Done

**Pull request merged**